lib/src/animated_sampler.dart

```diff
@@ -157,58 +154,54 @@ class _RenderShaderSamplerBuilderWidget extends RenderProxyBox {
     markNeedsPaint();
   }
 
-  ui.Image? _childRaster;
-
-  @override
-  void detach() {
-    _childRaster?.dispose();
-    _childRaster = null;
-    super.detach();
-  }
-
   @override
-  void dispose() {
-    _childRaster?.dispose();
-    _childRaster = null;
-    super.dispose();
-  }
+  bool get alwaysNeedsCompositing => true;
 
-  // Paint [child] with this painting context, then convert to a raster and detach all
-  // children from this layer.
-  ui.Image? _paintAndDetachToImage() {
-    final OffsetLayer offsetLayer = OffsetLayer();
-    final PaintingContext context =
-        PaintingContext(offsetLayer, Offset.zero & size);
-    super.paint(context, Offset.zero);
-    // This ignore is here because this method is protected by the `PaintingContext`. Adding a new
-    // method that performs the work of `_paintAndDetachToImage` would avoid the need for this, but
-    // that would conflict with our goals of minimizing painting context.
-    // ignore: invalid_use_of_protected_member
-    context.stopRecordingIfNeeded();
-    final ui.Image image = offsetLayer.toImageSync(Offset.zero & size,
-        pixelRatio: devicePixelRatio);
-    offsetLayer.dispose();
-    return image;
+  ui.Picture _paintBuilder(Offset offset, OffsetLayer childLayer) {
+    final ui.PictureRecorder pictureRecorder = ui.PictureRecorder();
+    final Canvas canvas = Canvas(pictureRecorder);
+    final ui.Image image = childLayer.toImageSync(
+      offset & size,
+      pixelRatio: devicePixelRatio,
+    );
+    try {
+      builder(image, size, offset, canvas);
+    } finally {
+      image.dispose();
+    }
+    return pictureRecorder.endRecording();
   }
 
-  @override
-  bool get alwaysNeedsCompositing => true;
-
   @override
   void paint(PaintingContext context, Offset offset) {
     if (size.isEmpty) {
-      _childRaster?.dispose();
-      _childRaster = null;
       return;
     }
 
     if (!enabled) {
-      _childRaster?.dispose();
-      _childRaster = null;
       return super.paint(context, offset);
     }
-    _childRaster?.dispose();
-    _childRaster = _paintAndDetachToImage();
-    builder(_childRaster!, size, offset, context.canvas);
+
+    context.pushLayer(
+      _ShaderSamplerBuilderLayer(offset, _paintBuilder),
+      (context, offset) {
+        context.pushLayer(OffsetLayer(), super.paint, offset);
+      },
```

Comment:
```suggestion
      super.paint,
```

I don't think you need this second layer.

Reply:
We cannot use `_ShaderSamplerBuilderLayer` as the layer that we call `toImageSync` on because `toImageSync` eventually calls `addToScene` on the target layer, which would lead to infinite recursion in this case.

Reply:
Yeah good, point - you'd need a different implementation of toImageSync that only added the children. I think this will work:

```dart
/// A [Layer] that uses an [AnimatedSamplerBuilder] to create a [ui.Picture]
/// every time it is added to a scene.
class _ShaderSamplerBuilderLayer extends OffsetLayer {
  _ShaderSamplerBuilderLayer(
    Offset offset,
    this.size,
    this.devicePixelRatio,
    this.callback,
  ) : super(offset: offset);

  final Size size;
  final double devicePixelRatio;
  final AnimatedSamplerBuilder callback;

  ui.Image _buildChildScene(Rect bounds, double pixelRatio) {
    final ui.SceneBuilder builder = ui.SceneBuilder();
    final Matrix4 transform = Matrix4.diagonal3Values(pixelRatio, pixelRatio, 1);
    transform.translate(-(bounds.left + offset.dx), -(bounds.top + offset.dy));
    builder.pushTransform(transform.storage);
    addChildrenToScene(builder);
    builder.pop();
    return builder.build().toImageSync(
      (pixelRatio * bounds.width).ceil(),
      (pixelRatio * bounds.height).ceil(),
    );
  }

  @override
  void addToScene(ui.SceneBuilder builder) {
    final ui.Image image = _buildChildScene(
      offset & size,
      devicePixelRatio,
    );
    final ui.PictureRecorder pictureRecorder = ui.PictureRecorder();
    final Canvas canvas = Canvas(pictureRecorder);

    try {
      callback(image, size, offset, canvas);
    } finally {
      image.dispose();
    }
    final ui.Picture picture = pictureRecorder.endRecording();
    builder.addPicture(Offset.zero, picture);
  }
}
```

Reply:
Nice! That works.